roachprod/roachtest: uniform storage capabilities

golgeek opened the pull request:
Until now, each cloud provider implementation had its own capabilities with regards to storage options. GCE was the only first class citizen with the most available options exposed in roachtest.

This patch attempts to bridge the feature parity gap between the cloud providers (up to what's exposed by each providers), bringing support for the following options in roachprod and roachtest:
- GCE:
  - local SSD
  - network disk size
  - network disk type (pd-standard, pd-ssd)
  - network disk count
  - RAID0 or multiple stores
- AWS:
  - local SSD
  - network disk size
  - network disk throughput
  - network disk IOPS
  - NEW: network disk type (gp2, gp3, io1, io2, st1, sc1, standard)
  - NEW: network disk count
  - NEW: RAID0 or multiple stores
- Azure:
  - local SSD
  - network disk size
  - NEW: network disk IOPS (ultra-disk only)
  - NEW: network disk type (standard-ssd, premium-ssd, premium-ssd-v2,
      ultra-disk)
  - NEW: network disk count
  - NEW: RAID0 or multiple stores
- IBM:
  - network disk size
  - network disk IOPS
  - network disk type (general-purpose, 5iops-tier, 10iops-tier, custom)
  - network disk count
  - RAID0 or multiple stores

This patch also splits the disk setup startup script snippets, with:
- a provider-specific way of detecting the attached disks
- a common logic to mount, format and aggregate the disks

This allows to offer the following filesystems across the board in roachprod (and roachtest):
- Ext4
- ZFS
- XFS
- F2FS (available for AWS, Azure and GCP, pending newer kernel for IBM)
- Btrfs

Epic: none
Closes: #123775
Informs: #146661, #113869
Release note: None

golgeek committed:
roachprod/roachtest: uniform storage capabilities

Until now, each cloud provider implementation had its own capabilities
with regards to storage options. GCE was the only first class citizen
with the most available options exposed in roachtest.

This patch attempts to bridge the feature parity gap between the cloud
providers (up to what's exposed by each providers), bringing support for
the following options in roachprod and roachtest:
- GCE:
  - local SSD
  - network disk size
  - network disk type (pd-standard, pd-ssd)
  - network disk count
  - RAID0 or multiple stores
AWS:
  - local SSD
  - network disk size
  - network disk throughput
  - network disk IOPS
  - NEW: network disk type (gp2, gp3, io1, io2, st1, sc1, standard)
  - NEW: network disk count
  - NEW: RAID0 or multiple stores
- Azure:
  - local SSD
  - network disk size
  - NEW: network disk IOPS (ultra-disk only)
  - NEW: network disk type (standard-ssd, premium-ssd, premium-ssd-v2,
    ultra-disk)
  - NEW: network disk count
  - NEW: RAID0 or multiple stores
- IBM:
  - network disk size
  - network disk IOPS
  - network disk type (general-purpose, 5iops-tier, 10iops-tier, custom)
  - network disk count
  - RAID0 or multiple stores

This patch also splits the disk setup startup script snippets, with:
- a provider-specific way of detecting the attached disks
- a common logic to mount, format and aggregate the disks

This allows to offer the following filesystems across the board in
roachprod (and roachtest):
- Ext4
- ZFS
- XFS
- F2FS (pending newer kernel for IBM)
- Btrfs

Epic: none
Closes: #123775
Informs: #146661, #113869
Release note: None

golgeek commented:
TC runs:
- [GCE](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/20706935)
- [AWS](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyAwsBazel/20706936)
- [Azure](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyAzureBazel/20706937)

DarrylWong commented on pkg/cmd/roachtest/spec/option.go:
```diff
@@ -248,17 +262,15 @@ func GCEMinCPUPlatform(platform string) Option {
 
 // GCEVolumeType sets the volume type when the cluster is on GCE.
+// Deprecated: use VolumeType instead.
 func GCEVolumeType(volumeType string) Option {
```
nit: I see 6 tests that use this deprecated option, can we just get rid of it now?

DarrylWong commented on pkg/roachprod/vm/startup_test.go:
```diff
@@ -33,8 +33,8 @@ func TestGenericStartupArgs(t *testing.T) {
 		require.Equal(t, args.DisksInitializedFile, "test")
 	})
 	t.Run("WithZfs", func(t *testing.T) {
```
nit: test name is now wrong

golgeek commented on pkg/roachprod/vm/startup.go:
```diff
@@ -331,6 +408,145 @@ const startupScriptUlimits = `
 # workers bump into this often.
 sudo sh -c 'echo "root - nofile 1048576\n* - nofile 1048576" > /etc/security/limits.d/10-roachprod-nofiles.conf'`
 
+const startupScriptMountDisks = `
```
Added a check to ensure we don't get anything crazy as a user input:
```
Error: UNCLASSIFIED_PROBLEM: unsupported filesystem: tmp
(1) UNCLASSIFIED_PROBLEM
Wraps: (2) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/roachprod/vm.ParseFileSystemOption
  |     pkg/roachprod/vm/vm.go:303
  | github.com/cockroachdb/cockroach/pkg/roachprod.Create
  |     pkg/roachprod/roachprod.go:1756
  | github.com/cockroachdb/cockroach/pkg/cmd/roachprod/cli.(*commandRegistry).buildCreateCmd.func1
  |     pkg/cmd/roachprod/cli/commands.go:126
  | github.com/cockroachdb/cockroach/pkg/cmd/roachprod/cli.(*commandRegistry).buildCreateCmd.Wrap.func3
  |     pkg/cmd/roachprod/cli/util.go:116
  | github.com/spf13/cobra.(*Command).execute
  |     external/com_github_spf13_cobra/command.go:920
  | github.com/spf13/cobra.(*Command).ExecuteC
  |     external/com_github_spf13_cobra/command.go:1044
  | github.com/spf13/cobra.(*Command).Execute
  |     external/com_github_spf13_cobra/command.go:968
  | main.main
  |     pkg/cmd/roachprod/main.go:43
  | runtime.main
  |     GOROOT/src/runtime/proc.go:285
  | runtime.goexit
  |     src/runtime/asm_arm64.s:1268
Wraps: (3) unsupported filesystem: tmp
Error types: (1) errors.Unclassified (2) *withstack.withStack (3) *errutil.leafError
```

DarrylWong commented on pkg/roachprod/roachprod.go:
```diff
@@ -1034,13 +1034,19 @@ func Reformat(ctx context.Context, l *logger.Logger, clusterName string, fs stri
 	switch fs {
-	case vm.Zfs:
-		if err := install.Install(ctx, l, c, []string{vm.Zfs}); err != nil {
+	case string(vm.Zfs):
```
Similar nit: would it be cleaner to just cast fs to a filesystem? `switch vm.Filesystem(fs) {`

golgeek committed:
roachprod: gce arm64 and local ssd machine types

Prior to this patch, the supported machine families in GCE was partial.
Only t2a was considered ARM64, and local SSD was only supported for n1
and n2 families.

This patch brings machine family parsing and capabilities deduction for
all machine types available in GCE as of today.

Epic: none
Release note: None
